### PR TITLE
Support ES2017 object rest/spread properties

### LINKED ISSUE #1588: Object Rest/Spread Properties
New productions:
```
PropertyDefinition[Yield] ::
  ...  AssignmentExpression[In, ?Yield]

ObjectAssignmentPattern[Yield] ::
  {AssignmentRestProperty[?Yield]opt}
  {AssignmentPropertyList[?Yield]  ,  AssignmentRestProperty[?Yield]opt}

AssignmentRestProperty[Yield] ::
  ...  DestructuringAssignmentTarget[?Yield]

ObjectBindingPattern[Yield] ::
  {BindingRestProperty[?Yield]opt}
  {BindingPropertyList[?Yield]  ,  BindingRestProperty[?Yield]opt}

BindingRestProperty[Yield] ::
  ...  BindingIdentifier[?Yield]
  ...  BindingPattern[?Yield]
```

AST changes:
```js
extend interface ObjectPattern {
    properties: [ AssignmentProperty | RestElement ];
}

extend interface ObjectExpression {
    properties: [ Property | SpreadElement ];
}
```

References:
- The proposal: http://sebmarkbage.github.io/ecmascript-rest-spread/
- Change to the spec: ?
- ESTree for ES2017: https://github.com/estree/estree/blob/master/es2017.md

### PR DESCRIPTION
**Pull request by ariya**
Fix #1588

### CODE REVIEW

File: src/messages.ts [hunk 1]
```diff
@@ -7,6 +7,7 @@ export const Messages = {
     ConstructorSpecialMethod: 'Class constructor may not be an accessor',
     DeclarationMissingInitializer: 'Missing initializer in %0 declaration',
     DefaultRestParameter: 'Unexpected token =',
+    DefaultRestProperty: 'Unexpected token =',
```
Comment (ariya) [inline src/messages.ts]: There is still no V8 error message for this, so we stick with a generic for the mean time.

File: src/messages.ts [hunk 2]
```diff
@@ -32,6 +33,7 @@ export const Messages = {
     NoAsAfterImportNamespace: 'Unexpected token',
     NoCatchOrFinally: 'Missing catch or finally after try',
     ParameterAfterRestParameter: 'Rest parameter must be last formal parameter',
+    PropertyAfterRestProperty: 'Unexpected token',
```
Comment (ariya) [inline src/messages.ts]: idem

**Comment by ariya** [comment]
One challenge: what if this feature doesn't make into ES2017 draft? According to https://github.com/tc39/proposals, it's still Stage 3 and there is a TC-39 meeting next week.

**Comment by mikesherov** [comment]
My vote would be to wait till the meeting before landing.

**Comment by ariya** [comment]
> My vote would be to wait till the meeting before landing.

On a second thought, I think we still have the choice of keeping it or reverting it in the 4.0 release branch (after TC-39 meeting).

**Comment by mikesherov** [comment]
Either way works! FWIW, I don't think this will end up like Object.observe. It's already heavily in use and is likely to make it to stage 4 eventually. So merge away if you'd like. I just was thinking it's cheap to wait till next week.

**Pull request closed without merging**